src/dash-table/handlers/cellEvents.ts

```diff
@@ -1,57 +1,69 @@
+import { makeCell, makeSelection } from 'dash-table/derived/cell/cellProps';
 import reconcile from 'dash-table/type/reconcile';
 
-function isCellSelected(selectedCells: SelectedCells, idx: number, i: number) {
-    return selectedCells && R.contains([idx, i], selectedCells);
-}
-
 export const handleClick = (propsFn: () => ICellFactoryProps, idx: number, i: number, e: any) => {
     const {
         selected_cells,
+        start_cell,
         setProps,
-        virtualized
+        virtualized,
+        columns,
+        viewport
     } = propsFn();
 
-    const selected = isCellSelected(selected_cells, idx, i);
+    const selected = isSelected(selected_cells, idx, i);
 
     // don't update if already selected
     if (selected) {
         return;
     }
 
     e.preventDefault();
-    const cellLocation: [number, number] = [
-        idx + virtualized.offset.rows,
-        i + virtualized.offset.columns
-    ];
+
+    const row = idx + virtualized.offset.rows;
+    const col = i + virtualized.offset.columns;
+
+    const activeCell = makeCell(row, col, columns, viewport);
 
     const newProps: Partial<ICellFactoryProps> = {
         is_focused: false,
-        active_cell: cellLocation
+        active_cell: activeCell,
+        start_cell: activeCell,
+        end_cell: activeCell
     };
 
-    const selectedRows = R.uniq(R.pluck(0, selected_cells)).sort((a, b) => a - b);
-    const selectedCols = R.uniq(R.pluck(1, selected_cells)).sort((a, b) => a - b);
-    const minRow = selectedRows[0];
-    const minCol = selectedCols[0];
-
-    if (e.shiftKey) {
-        newProps.selected_cells = R.xprod(
-            R.range(
-                R.min(minRow, cellLocation[0]),
-                R.max(minRow, cellLocation[0]) + 1
-            ),
-            R.range(
-                R.min(minCol, cellLocation[1]),
-                R.max(minCol, cellLocation[1]) + 1
-            )
-        ) as any;
+    if (e.shiftKey && start_cell && start_cell.row !== undefined) {
+        newProps.selected_cells = makeSelection(
+            {
+                minRow: min(row, start_cell.row),
+                maxRow: max(row, start_cell.row),
+                minCol: min(col, start_cell.column),
+                maxCol: max(col, start_cell.column)
+            },
+            columns,
+            viewport
+        );
+        delete newProps.start_cell;
+        delete newProps.active_cell;
+
     } else {
-        newProps.selected_cells = [cellLocation];
+        newProps.selected_cells = [activeCell];
     }
 
+    /*
+     * In some cases this will initiate browser text selection.
+     * We've hijacked copying, so while it might be nice to allow copying part
+     * of a cell, currently you'll always get the whole cell regardless of what
+     * the browser thinks is selected.
+     * And when you've selected multiple cells the browser selection is
+     * completely wrong.
+     */
+    window.getSelection().removeAllRanges();
```

Comment:
For example, shift-click to select in the first example at https://dash.plot.ly/datatable:
<img width="845" alt="Screen Shot 2019-04-22 at 1 43 56 AM" src="https://user-images.githubusercontent.com/2678795/56485004-377d1500-64a0-11e9-90bc-644006c02e0e.png">